cri: add tracing spans for image pull and sandbox setup paths

Cindia-blue opened the pull request:
This PR adds opt-in tracing spans/attributes in CRI image pull and selected sandbox-related paths to improve debugging and correlation (e.g., sandbox.id/pod metadata). If maintainers prefer a smaller diff, I’m happy to split this into a pull-only PR plus follow-ups.

Signed-off-by: Cindy Li <xinhuil@netflix.com>

mikebrow reviewed (commented):
we already have spans coving the CRI calls.. could you explain what is broken?

Cindia-blue commented:
> we already have spans coving the CRI calls.. could you explain what is broken?

Thanks Mike — totally agree that we already have spans covering the CRI gRPC calls.

What’s “broken” for us is that the existing RPC spans don’t provide enough visibility/correlation into the *work done underneath* the CRI call boundaries, especially when failures happen in the middle of pull/unpack/registry interactions (or when we need to distinguish *which stage* is slow/failed). Concretely:

- The CRI PullImage RPC span alone doesn’t tell us where time was spent (resolve vs dispatch vs convert manifest vs unpack), nor does it reliably carry the identifiers we need to correlate across layers (e.g., sandbox/pod context when available).
- When we see CreateContainer/StartContainer issues during cold-start style scenarios, the “RPC is slow/failed” signal is not sufficient to pinpoint whether the bottleneck is registry resolution, content dispatch, unpack, or subsequent setup steps.
- The intent of this PR is to add opt-in, lightweight in-process spans/attributes around those internal stages so a single trace can answer “which step broke/was slow” without relying on parsing logs.

If you think this belongs strictly in the existing CRI RPC spans (e.g., via additional events/attributes) rather than new child spans in the underlying paths, I’m happy to adjust — would appreciate your guidance on the preferred pattern.

Cindia-blue commented:
Hi @fuweid, @estesp, @dmcgowan, @mikebrow

I’ve put together a PR that adds **minimal, opt-in tracing spans** along the CRI image pull path and a few closely related runtime/sandbox flows.

The intent is to complement existing CRI gRPC tracing with lightweight, in-process spans (e.g. resolve / dispatch / unpack / sandbox setup), to improve correlation and debugging without changing execution semantics or introducing mandatory overhead.

CI is green, and I’ve tried to keep the scope focused and the changes self-contained. I’d really appreciate your thoughts on:
- whether the scope feels appropriate for containerd,
- and whether this direction aligns with how tracing is expected to evolve in the project.

If the diff still feels too broad, I’m happy to split this into a smaller pull-only PR and follow up separately.

Thanks for taking a look — feedback very welcome.

samuelkarp commented:
> If the diff still feels too broad, I’m happy to split this into a smaller pull-only PR and follow up separately.

I would recommend this; it'll help to have smaller single-purpose diffs.

> This branch is out-of-date with the base branch

It'll need a rebase too.

Cindia-blue commented:
Thanks @samuelkarp done.

I rebased onto latest main and split the PR. This PR is now pull-only (client pull + CRI image pull). Follow-up changes (task/metadata/sandbox/CNI spans) will come as separate PR(s).

mikebrow commented on internal/cri/server/images/image_pull.go:
```diff
@@ -97,12 +97,65 @@ import (
+		// Common fallback keys that may carry sandbox id.
+		if v, ok := r.SandboxConfig.Annotations["io.kubernetes.cri.sandbox-id"]; ok && v != "" {
+			sandboxID = v
+		} else if v, ok := r.SandboxConfig.Annotations["k8s.v1.cni.cncf.io/sandbox-id"]; ok && v != "" {
```
??

mikebrow replied:
thx don't get me wrong one day we might do an OCI/CNCF pod/sandbox specification .. just not yet.. not completely surprised the container networking folks (cni) had an annotation given the default network namespace being attached to all containers in the sandbox.

mikebrow reviewed (commented):
couple comments..

mikebrow commented on internal/cri/server/images/image_pull.go:
```diff
@@ -117,15 +165,42 @@ func (c *GRPCCRIImageService) PullImage(ctx context.Context, r *runtime.PullImag
 
 func (c *CRIImageService) PullImage(ctx context.Context, name string, credentials func(string) (string, string, error), sandboxConfig *runtime.PodSandboxConfig, runtimeHandler string) (_ string, err error) {
-	span := tracing.SpanFromContext(ctx)
+	ctx, span := tracing.StartSpan(ctx, tracing.Name("cri.image", "pull"))
```
probably don't need a new span here..  I believe it would be more important to make sure the paths that get here have already set the span in the passed in ctx.  

Example.. run pod sandbox creating a core internal sandboxer/sandbox will ensure the sandbox pause container image is local..

xinhuil-max committed:
client: trace image pull stages

Signed-off-by: Cindy Li <xinhuil@netflix.com>

Cindia-blue commented:
Hi @fuweid I’ve refreshed this PR on the latest main and kept the change small and single-purpose. When you have a chance, could you please take a look?

Cindia-blue commented:
Hi @mikebrow — thanks for the update on Aug 3. CI is now green. Is there anything else you’d like me to address before this is ready for approval or merge?
